rfcs/text/0014_api_documentation.md

```diff
@@ -0,0 +1,439 @@
+
+Core has a pattern of writing comments like this:
+
+```ts
+  /** {@link IUiSettingsClient} */
+  uiSettings: IUiSettingsClient;
+```
+
+I don't see the value in this. In the IDE, I can click on the IUiSettingsClient type and get directed there, and in the API doc system, the
+type will already be clickable. This ends up with a weird looking API:
+
+![image](../images/repeat_type_links.png)
+
+The plan is to make @link comments work like links, which means this is unneccessary information.
+
+I propose we avoid this kind of pattern.
+
+## Export every referenced type
+
+The docs system handles broken link warnings but to avoid breaking the ci, I suggest we turn this off initially. This will mean however,
+that we may miss situations where we are referencing a type that is not actually exported. This will cause a broken link in the docs
+system
+
+For example if your index.ts file has:
+```ts
+export type foo: string | AnInterface;
+```
+
+and does not also export `AnInterface`, this will be a broken link in the docs system.
+
+Until we have better ci tools to catch these mistakes, developers will need to export every referenced type.
+
+## Avoid `Pick` pattern
+
+Connected to the above, if you use `Pick`, there are two problems. One is that it's difficult for a developer to see the functionality
+available to them at a glance, since they would have to keep flipping from the interface definition to the properties that have been picked.
+
+The second potential problem is that you will have to export the referenced type, and in some situations, it's an internal type that isn't exported.
+
+![image](../images/api_doc_pick.png)
+
+# Open questions
+
+## Required attribute
+
+`isRequired` is an optional parameter that can be used to display a badge next to the API.
+We can mark function parameters that do not use `?` or `| undefined` as required. Open questions:
+
+1. Are we okay with a badge showing for `required` rather than `optional` when marking a parameter as optional is extra work for a developer, and `required` is the default?
+
+2. Should we only mark function parameters as `required` or interface/class parameters? Essentially, should any declaration that is not nullable
+have the `required` tag?
+
+## Signatures on primitive types
+
+1. Should we _always_ include a signature for variables and parameters, even if they are a repeat of the TypeKind? For example:
+
+![image](../images/repeat_primitive_signature.png)
+
+2. If no, should we include signatures when the only difference is `| undefined`? For function parameters this information is captured by
+the absence of the `required` badge. Is this obvious? What about class members/interface props?
+
+## Out of scope
+
+### REST API
+
+This RFC does not cover REST API documentation, though it worth considering where
+REST APIs registered by plugins should go in the docs. The docs team has a proposal for this but it is not inside the `Kibana Developer Docs` mission.
```

Comment:
Plugins shouldn't be consuming other plugin's HTTP APIs directly, right? Instead, it should be plugin to plugin communication, which might then go over HTTP...

Reply:
I don't think we have that written down as a recommendation anywhere.  I believe there are some cases where there is no plugin API so devs go through the HTTP API (I think some may do this with the short URL feature, or at least ML was looking into it).

Either way, I think you are saying they shouldn't go in these developer docs? They are for developers... but not necessarily developers writing Kibana plugins. So I think I'd be on board with that.

Reply:
> Either way, I think you are saying they shouldn't go in these developer docs? They are for developers... but not necessarily developers writing Kibana plugins. So I think I'd be on board with that.

Yeah, they're potentially for developers, but they're also for the general public to consume as well.